cmd/loadgenerator/core.go

```diff
@@ -0,0 +1,74 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type loadGenerator struct {
+	cfg    Config
+	now    func() time.Time
+	logger *logp.Logger
+}
+
+func (l loadGenerator) Run(ctx context.Context) error {
+	t := l.now()
+	delta := l.cfg.TimeDelta
+	delay := l.cfg.Delay
+
+	// Loop forever
+	if l.cfg.Loop {
+		for {
+			select {
+			case <-ctx.Done():
+				l.logger.Info("Shutdown signal received, closing loadGenerator")
+				return nil
+			default:
+				if err := l.send(l.next(t)); err != nil {
+					return fmt.Errorf("error sending event: %w", err)
+				}
+				if l.cfg.CurrentTime {
+					t = l.now()
+				} else {
+					t = t.Add(delta)
+				}
+				time.Sleep(delay)
+			}
+		}
+	}
+
+	// Generate a set number of events
+	for i := uint64(0); i < l.cfg.EventsCount; i++ {
+		if err := l.send(l.next(t)); err != nil {
+			return fmt.Errorf("error sending event: %w", err)
+		}
+		if l.cfg.CurrentTime {
+			t = l.now()
+		} else {
+			t = t.Add(delta)
+		}
+		time.Sleep(delay)
+	}
+
+	return nil
+}
+
+// next returns the next event
+func (l loadGenerator) next(t time.Time) string {
```

Comment:
I'd recommend `[]byte` as the return type, that lines up better for the outputs which will usually use a `Write` interface to output.  Also be nice if these were interfaces, maybe:

```
type Output interface {
	Write(p []byte) (n int, err error)
	Close() error
}

type Generator interface {
	Next() (p []byte, err error)
}
```

Reply:
My idea was to postpone most of thsoe decisions to when I start integration with the publishing pipeline and shipper.

The current shipper implementation (and client) use gRPC calls that accept a Protobuf message that does not have a representation to `[]byte` ([event type](https://github.com/elastic/elastic-agent-shipper-client/blob/ce5fac5d4177ece459766ff56d39028ed28da937/api/messages/publish.proto#L32-L43) and [stuct/JSON type](https://github.com/elastic/elastic-agent-shipper-client/blob/ce5fac5d4177ece459766ff56d39028ed28da937/api/messages/struct.proto#L46-L75)), so I believe it makes sense to keep the current implementation.

There are already some abstractions from the [input manager inherited from Beats](https://github.com/elastic/elastic-agent-inputs/blob/5c7785eb0e735a475e7a4ddabc1cea89ee83789e/pkg/manager/input/input.go#L48-L66). Once we start integrating the code with the input manager, the necessary interfaces will became more clear.